### PR TITLE
Added animation to save icon when user enters a comment text

### LINKED ISSUE #9362: Animate Save Icon When User Enters Comment Text
## `Description`
From a user interaction standpoint, it would be helpful to have the save icon blink whenever a user enters text into a comment form.

The save icon is the 💾 in the comment toolbar below:

Screen Shot 2021-03-22 at 7 11 24 PM

See #9131, particularly [this comment](https://github.com/publiclab/plots2/issues/9131#issuecomment-776245431).

Currently the comment form saves every single time the input is changed. This isn't very useful to the user, so we'll also want to implement some kind of [debouncing](https://www.freecodecamp.org/news/javascript-debounce-example/) or `setTimeout` delay.

## `Summary of Changes to Be Made`
These changes can be done in separate PRs, and by different contributors. But they probably have to be done in sequence:

1. Add debouncing to delay the `save` method in `editor.js`, either wait 1000ms, or only run `save` when user input has stopped for 500ms (you choose the time).
2. Add some animation to the 💾 save icon when the `save` method runs, so that user knows that it's running. Darken the icon background. Replace the 💾 with one of [these FontAwesome spinner icons](https://fontawesome.com/how-to-use/on-the-web/styling/animating-icons).
3. Swap out the `Drag and drop to add an image or file, or choose one` with `Saving...` to absolutely spell it out for the user.

## `Useful Links & Images`
What the end goal looks like:
Screen Shot 2021-03-23 at 12 04 39 PM

These changes are going to be made in the comment/legacy editor, which lives at `app/assets/javascripts/editor.js`:
https://github.com/publiclab/plots2/blob/62fdcd0dd54b47727a046a0129ba138710082f0b/app/assets/javascripts/editor.js#L116-L119

Leave a comment below if you want to claim this issue, and post here if you need any help getting oriented to the codebase! Good luck! 🍀

### PR DESCRIPTION
**Pull request by waridrox**
Fixes #9362 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

Final working version - 

https://user-images.githubusercontent.com/58583793/112806713-b7508880-9094-11eb-9d2a-76de29f88b53.mp4

_I've removed the `console.log` statements 😄._ 

@noi5e kindly review this. Thanks :)

### CODE REVIEW

**Comment by waridrox** [comment]
> One thing to note is that the save & recover system test needs to be updated (that's what's failing right now).

Oh!😅 that's what is failing...

**Comment by waridrox** [comment]
> > One thing to note is that the save & recover system test needs to be updated (that's what's failing right now).
> 
> Oh!😅 Will take a look at this soon...

**Comment by waridrox** [comment]
My apologies, got a little busy 😅, @noi5e can you point me to the right file to implement the tests for debounce function. Further researching upon, I came across the `sleep` method in ruby which waits the program for a certain amount of time.

**Comment by RuthNjeri** [comment]
Hi @waridrox, from the system tests [here](https://github.com/publiclab/plots2/pull/9384/checks?check_run_id=2216654085) the failing test is 
Screenshot 2021-04-15 at 22 14 49

`comment_test.rb`

**Comment by waridrox** [comment]
Just when I though about adding tests, I realised we missed an edge case wherein due to the similar functional elements like the toolbar and the image upload section, the current changes also reflect if a `Reply to a Comment` is open along with `Post a Comment` window. Here's an example for that case and I'm working on first fixing that! 


https://user-images.githubusercontent.com/58583793/115920687-2a80cb00-a498-11eb-9b46-9b3cba1d9f1c.mp4

**Comment by waridrox** [comment]
Here's the fix for the edge case: reason being that each comment has its own unique id associated so earlier code was targeting all ids instead of the specific one that the user is currently making changes on.


https://user-images.githubusercontent.com/58583793/116014772-927c1080-a654-11eb-97e8-14bf5da79322.mp4


@noi5e could you give some insights on testing the new changes since there are now ids associated with each comment section 😅 Thanks :)

**Comment by waridrox** [comment]
The tests have been updated to account for debounce function and tests for save and recover button pass now :)
Screenshot 2021-04-27 at 6 11 41 PM

File: app/assets/javascripts/editorToolbar.js [hunk 1]
```diff
@@ -90,9 +90,52 @@ $(function() {
     })
     // for save & recover buttons
     .on("click", ".save-button", function(e) {
-      $E.setState(e.currentTarget.dataset.formId); // string that is: "main", "reply-123", "edit-123" etc.
-      $E.save($E);
-    })
+      //preventing multiple clicks on the button
+      if(!e.detail || e.detail == 1) {
```
[the remark this answers is not in the document]
Comment (waridrox) [reply app/assets/javascripts/editorToolbar.js:94]: Isn't `===`  supposed to check type as well 🤔. I think here `==` should suffice...

**Comment by RuthNjeri** [comment]
Hi @waridrox, what do you think about the other comments made by CodeClimate? Do you think you can address them? I see you responded to one...would you like to address the rest? We usually don't implement everything suggested by CodeClimate depending on the trade offs.

**Comment by waridrox** [comment]
> Hi @waridrox, what do you think about the other comments made by CodeClimate? Do you think you can address them? I see you responded to one...would you like to address the rest? We usually don't implement everything suggested by CodeClimate depending on the trade offs.

Yes, @RuthNjeri, I've looked at the comments by CodeClimate, most of them being a repetition of one another 😅. But I don't see any problem in the issues that it is suggesting to be fixed - 

- Function attachSaveListener has 32 lines of code (exceeds 25 allowed). - This function also houses comments for explanations, but CodeClimate is also counting those in the number of lines.
- Similar blocks of code found in 2 locations. Consider refactoring. - The code do start similar but eventually do different things at the completion.
- Expected '===' and instead saw '=='. - It means that '===' is being enforced unnecessarily when '==' can take of that.

**Comment by waridrox** [comment]
@noi5e @RuthNjeri @jywarren, pls review the same, this seems complete from my end. Thanks :)

**Comment by waridrox** [comment]
@noi5e @jywarren pls review the changes. Thanks :)

File: test/system/comment_test.rb [hunk 2]
```diff
@@ -359,6 +363,8 @@ def get_path(page_type, path)
       page.find('#text-input-edit-' + edit_id_num)
         .click
         .fill_in with: comment_text_edit
+        # taking into account the time for debounce function
+        sleep(0.7)
```
Comment (RuthNjeri) [inline test/system/comment_test.rb:367]: @waridrox, what is the purpose of the sleep? Will the tests fail if the sleep is removed? And why 0.7 seconds?

File: app/assets/javascripts/editorToolbar.js [hunk 3]
```diff
@@ -90,9 +90,52 @@ $(function() {
...
+      if(!e.detail || e.detail === 1) {
+          //explicitly handling main comment section
+          if ((this.id) === "save-button-main") {
+            // toggling the favicon save icon class to add a spinner icon
+            $(this).find("i").toggleClass("fa fa-save fas fa-sync fa-spin");
+      
+            //changing the text from "Upload an image" to "Saving..."
+            let saving_text = $('<p id="saving-text"> Saving... </p>');
+            $("#comment-form-main .imagebar").prepend(saving_text);
+            $("#comment-form-main .imagebar p").not("#saving-text").hide();
```
Comment (RuthNjeri) [inline app/assets/javascripts/editorToolbar.js:103]: Is this change of text to make it re-usable with the comment save? Meaning that if it stays `upload an image` it would mean that this text would appear when saving a text?

File: test/system/comment_test.rb (hunk 2 above)
Comment (waridrox) [reply test/system/comment_test.rb:367]: @RuthNjeri, here the purpose of `sleep` is to wait the tests for 0.7 seconds in order for the save function to take effect due to the debounce function, whenever new content is put in the comment text area: 
```javascript
    }, 700)
     );
   }  
     save(thisEditor) {
```

The tests would fail because this would be a new addition to the comment area hence the use of `sleep` is to update the tests.

File: app/assets/javascripts/editorToolbar.js (hunk 3 above)
Comment (waridrox) [reply app/assets/javascripts/editorToolbar.js:103]: Yes, indeed! The text `upload an image` would be hidden and in turn the text `Saving...` would be displayed to the user at the image upload text section of the comment form.

Comment (waridrox) [reply app/assets/javascripts/editorToolbar.js:103]: The `upload an image` text is simply not interchanged with the `Saving...` text because the `upload an image` text also houses a hidden text, i.e. `Uploading` when an image is dragged on the section...

File: app/assets/javascripts/editor.js [hunk 4]
```diff
@@ -104,16 +104,74 @@ class Editor {
...
+      var context = this,
+        args = arguments;
+      var later = function () {
+        timeout = null;
+        if (!immediate) func.apply(context, args);
+      };
+      var callNow = immediate && !timeout;
+      clearTimeout(timeout);
+      timeout = setTimeout(later, wait);
+      if (callNow) func.apply(context, args);
```
Comment (RuthNjeri) [inline app/assets/javascripts/editor.js:121]: I have less experience using JS than Rails, is there a reason for using var over the latest let here?

Comment (waridrox) [reply app/assets/javascripts/editor.js:121]: `var` is used for global scope but I guess here it doesn't matter since the values it holds are accessed only by the debounce function. Thanks for the catch!!!

File: app/assets/javascripts/editor.js [hunk 5]
```diff
@@ -104,16 +104,74 @@ class Editor {
...
+    let timeout;
+    return function () {
+      let context = this,
+        args = arguments;
+      let later = function () {
+        timeout = null;
+        if (!immediate) func.apply(context, args);
+      };
+      let callNow = immediate && !timeout;
+      clearTimeout(timeout);
```
Comment (RuthNjeri) [inline app/assets/javascripts/editor.js:119]: Why, clear timeout here, if the variable has a null value in line 115, why not just reassign it on line 120 and skip doing line 119?

File: app/assets/javascripts/editorToolbar.js (hunk 3 above)
Comment (RuthNjeri) [reply app/assets/javascripts/editorToolbar.js:103]: Why not have a specific class or id for the text functionality and then hide the imagebar class/text? Is it easier to just reuse the imagebar class?

File: app/assets/javascripts/editor.js (hunk 5 above)
Comment (waridrox) [reply app/assets/javascripts/editor.js:119]: @RuthNjeri, at line 115 - 
```javascript
timeout = null;
```
the timeout variable gets cleared to make the next execution run as soon as the function is called, i.e. at line 120:
```javascript
      if (callNow) func.apply(context, args);
```

On the other hand, `clearTimeout` at line 119:
```javascript
clearTimeout(timeout);
```
 is used to clear out the previous created timeout, so clearTimeout should be done before setTimeout at line 120: 
```javascript
      timeout = setTimeout(later, wait);
```
in order to cancel out previous invocation of setTimeout function. Thus, `clearTimeout` is essential for the invocation the the `setTimeout` function 😅...

File: app/assets/javascripts/editorToolbar.js (hunk 3 above)
Comment (waridrox) [reply app/assets/javascripts/editorToolbar.js:103]: That may involve refactoring text functionality class code / styles associated with the current imagebar id in different places...That's why imagebar class was reused 😅

**Comment by waridrox** [comment]
Sorry the tests did not run and instead failed...

**Comment by pydevsg** [comment]
Refactor the code @waridrox. 
This will pass the test 😅

**Comment by jywarren** [comment]
If you'd like we can bypass the code climate suggestions! They are only suggestions but if you'd like to address any of them we can also wait, no rush!

**Comment by waridrox** [comment]
@jywarren @pydevsg, I was talking about all the tests instead of just the code climate ones😅I've tried to address them but all of them seem fine to me...

**Review by RuthNjeri — APPROVED** [review]
Thanks for working on this @waridrox, it looks good to me, I think one more reviewer should approve this(@jywarren or @cesswairimu)...then it will be okay to merge 🚀

**Comment by waridrox** [comment]
After updating the branch with upstream, unit tests seem to be failing...🤔

**Comment by jywarren** [comment]
Aha - for some reason i also saw this on https://github.com/publiclab/plots2/pull/9665#issuecomment-848231223 and suggested a fix there - adding a sort to the test or to the main codebase, as i believe it's an ordering issue that can happen sometimes if we don't explicitly send a sort to the database. If you're able to open a new PR then possibly this and https://github.com/publiclab/plots2/pull/9665#issuecomment-848231223 can both be rebased over it once we resolve that. In the meantime, i'll restart the test to see if we get past it that way. Thanks!

**Comment by jywarren** [comment]
Ok sorry @waridrox but now the unit tests upstream are fixed and just one more rebase should work. Many thanks for your patience!!!

**Comment by waridrox** [comment]
@jywarren Done!

**Comment by jywarren** [comment]
OK, merging now, and many thanks!!!

Regarding the revert button, i wonder if we should have saving add the text to an array of recent texts that have been saved. Then revert could just cycle back through that list.

**Comment by jywarren** [comment]
@waridrox @noi5e we could break out that idea of the array of comments saves into a new issue too! I think we'd have to be careful to flush the list... maybe only save 5 or 10 states, and/or delete much older ones or something? For example, what if someone types in something private, it saves, and they change their mind before posting it, they log out, someone else logs in, and they are able to recover the text.

**Comment by jywarren** [comment]
i.e. we could make it auto-delete texts older than an hour to mitigate this risk maybe.

**Comment by jywarren** [comment]
I know @waridrox is now working on other projects, though! So perhaps someone else from the @publiclab/soc team would be interested in taking this idea of fixing the comment revert button up?
